File: services/restlet.py
import os
from typing import Coroutine, Callable, Tuple, Awaitable
from urllib.parse import urlencode

import aiohttp
import oauthlib.oauth1

OAUTH_CLIENT = oauthlib.oauth1.Client(
    client_key=os.getenv("CONSUMER_KEY"),
    client_secret=os.getenv("CONSUMER_SECRET"),
    resource_owner_key=os.getenv("ACCESS_TOKEN"),
    resource_owner_secret=os.getenv("TOKEN_SECRET"),
    realm=os.getenv("ACCOUNT_ID"),
    signature_method=oauthlib.oauth1.SIGNATURE_HMAC_SHA256,
)

BASE_URL = f"https://{os.getenv('ACCOUNT_ID')}.restlets.api.netsuite.com/app/site/hosting/restlet.nl"


def request_restlet(restlet: dict, method: str) -> Callable:
    async def request(
        session: aiohttp.ClientSession,
        params: dict,
        body: dict,
    ) -> tuple[Exception | None, dict | None]:
        url, headers, body = OAUTH_CLIENT.sign(
            uri=f"{BASE_URL}?{urlencode({**restlet, **params})}",
            http_method=method,
            body=body,
            headers={
                "Content-type": "application/json",
            },
        )
        try:
            async with session.request(
                method,
                url,
                headers=headers,
            ) as r:
                res = await r.json()
            return None, res
        except Exception as e:
            print(e)
            return e, None

    return request

File: src/libs/restlet.ts
import * as dotenv from 'dotenv';
import axios from 'axios';
import type { Restlet, RestletMethod } from '../types/restlet';

dotenv.config();

const crypto = require('crypto');
const OAuth = require('oauth-1.0a');

const BASE_URL = `https://${process.env.ACCOUNT_ID}.restlets.api.netsuite.com/app/site/hosting/restlet.nl`;

const oauth = OAuth({
  consumer: {
    key: process.env.CONSUMER_KEY,
    secret: process.env.CONSUMER_SECRET,
  },
  realm: process.env.ACCOUNT_ID,
  signature_method: 'HMAC-SHA256',
  hash_function(baseString: string, key: any) {
    return crypto.createHmac('sha256', key).update(baseString).digest('base64');
  },
});

const axClient = axios.create();

const requestRestlet = (restlet: Restlet, method: RestletMethod) => {
  const request = async ({ params, body }: { params?: any; body?: any }) => {
    const requestParams = new URLSearchParams({
      script: restlet.script,
      deploy: restlet.deploy,
      ...params,
    });
    const requestURL = `${BASE_URL}?${requestParams.toString()}`;
    const headers = oauth.toHeader(
      oauth.authorize(
        {
          method,
          url: requestURL,
          data: body,
        },
        {
          key: process.env.ACCESS_TOKEN,
          secret: process.env.TOKEN_SECRET,
        }
      )
    );
    try {
      const { data } = await axClient.request({
        method,
        headers,
        url: requestURL,
        data: body,
      });
      return [null, data];
    } catch (err) {
      return [err, null];
    }
  };
  return request;
};

export default requestRestlet;
